docs/usage/types.md

```diff
@@ -0,0 +1,648 @@
+
+**`as_hsl_tuple`**
+: returns a 3- or 4-tuple in HSL(a) format, the `alpha` keyword argument can be used to define whether
+  the alpha channel should be included,
+  options: `True` - always include, `False` - never include, `None` (the default)  - include if set
+
+The `__str__` method for `Color` returns `self.as_named(fallback=True)`.
+
+!!! note
+    the `as_hsl*` refer to hue, saturation, lightness "HSL" as used in html and most of the world, **not**
+    "HLS" as used in python's `colorsys`.
+
+### Secret Types
+
+You can use the `SecretStr` and the `SecretBytes` data types for storing sensitive information
+that you do not want to be visible in logging or tracebacks.
+The SecretStr and SecretBytes will be formatted as either `'**********'` or `''` on conversion to json.
+
+```py
+{!./examples/ex_secret_types.py!}
+```
+_(This script is complete, it should run "as is")_
+
+### Json Type
+
+You can use `Json` data type - *Pydantic* will first parse raw JSON string and then will validate parsed object
+against defined Json structure if it's provided.
+
+```py
+{!./examples/ex_json_type.py!}
+```
+_(This script is complete, it should run "as is")_
+
+### Payment Card Numbers
+
+The `PaymentCardNumber` type validates [payment cards](https://en.wikipedia.org/wiki/Payment_card)
+(such as a debit or credit card).
+
+```py
+{!./examples/payment_card_number.py!}
+```
+_(This script is complete, it should be run "as is")_
+
+`PaymentCardBrand` can be one of the following based on the BIN:
+
+* `PaymentCardBrand.amex`
+* `PaymentCardBrand.mastercard`
+* `PaymentCardBrand.visa`
+* `PaymentCardBrand.other`
+
+The actual validation verifies the card number is:
+
+* a `str` of only digits
+* [luhn](https://en.wikipedia.org/wiki/Luhn_algorithm) valid
+* the correct length based on the BIN, if Amex, Mastercard or Visa, and between
+  12 and 19 digits for all other brands
+
+## Constrained Types
+
+The value of numerous common types can be restricted using `con*` type methods:
+
+```py
+{!./examples/constrained_types.py!}
+```
+_(This script is complete, it should run "as is")_
+
+## Strict Types
+
+You can use the `StrictStr`, `StrictInt`, `StrictFloat`, and `StrictBool` types
+to prevent coercion from compatible types.
+These types will only pass validation when the validated value is of the respective type or is a subtype of that type.
+This behavior is also exposed via the `strict` field of the `ConstrainedStr`, `ConstrainedFloat` and
+`ConstrainedInt` classes and can be combined with a multitude of complex validation rules.
+
+The following caveats apply:
+
+- `StrictInt` (and the `strict` option of `ConstrainedInt`) will not accept `bool` types,
+    even though `bool` is a subclass of `int` in Python. Other subclasses will work.
+- `StrictFloat` (and the `strict` option of `ConstrainedFloat`) will not accept `int`.
+
+```py
+{!./examples/strict_types.py!}
+```
+_(This script is complete, it should run "as is")_
+
+## Custom Data Types
```

Comment:
For what it's worth, I personally think this section, short as it is, merits its own top level page, separate from all the other types.

(It may also be worth expanding that page..)

This is such a useful feature and I think it doesn't get the attention it deserves -- it's a shame for it to be buried way at the bottom of the long list of types pydantic provides.

Reply:
okay, personally I think it's right here but we should link to it more and expand it to contain more details and a better example. Certainly the current example is awful.